File: cloud/mdb/dbaas-internal-api-image/dbaas_internal_api/utils/maintenance.py
"""
Common maintenance operations
"""

from datetime import datetime
from typing import Optional

from ..core.exceptions import DbaasClientError, NoChangesError
from .time import now
from ..utils import metadb
from ..utils.types import ClusterInfo, MaintenanceRescheduleType
from ..utils.maintenance_helpers import calculate_nearest_maintenance_window


class MaxDelayExceeded(DbaasClientError):
    pass


def reschedule_maintenance(
    cluster: ClusterInfo,
    reschedule_type: MaintenanceRescheduleType,
    delayed_until: Optional[datetime],
) -> datetime:
    if cluster.planned_operation is None:
        raise DbaasClientError('There is no maintenance operation at this time')

    new_maintenance_time = cluster.planned_operation.delayed_until
    if reschedule_type == MaintenanceRescheduleType.unspecified:
        raise DbaasClientError('Reschedule type need to be specified')
    elif reschedule_type == MaintenanceRescheduleType.immediate:
        new_maintenance_time = now()
    elif reschedule_type == MaintenanceRescheduleType.specific_time:
        if delayed_until is None:
            raise DbaasClientError('Delayed until time need to be specified with specific time rescheduling type')
        new_maintenance_time = delayed_until
    elif reschedule_type == MaintenanceRescheduleType.next_available_window:
        if cluster.maintenance_window.is_anytime:
            raise DbaasClientError('The window need to be specified with next available window rescheduling type')
        new_maintenance_time = calculate_nearest_maintenance_window(
            cluster.planned_operation.delayed_until,
            cluster.maintenance_window.weekly_maintenance_window.day,  # type: ignore
            cluster.maintenance_window.weekly_maintenance_window.hour,  # type: ignore
        )
    if new_maintenance_time == cluster.planned_operation.delayed_until:
        raise NoChangesError()

    if (
        new_maintenance_time > cluster.planned_operation.latest_maintenance_time
        and reschedule_type != MaintenanceRescheduleType.immediate
    ):
        raise MaxDelayExceeded(
            'Maintenance operation time cannot exceed {max_delay}'.format(
                max_delay=cluster.planned_operation.latest_maintenance_time
            )
        )

    metadb.reschedule_maintenance_task(cluster.cid, cluster.planned_operation.config_id, new_maintenance_time)
    return new_maintenance_time

File: cloud/mdb/mdb-internal-api/internal/logic/internal/clusters/provider/maintenance.go
package provider

import (
	"fmt"
	"strings"
	"time"

	"a.yandex-team.ru/cloud/mdb/internal/optional"
	"a.yandex-team.ru/cloud/mdb/internal/semerr"
	"a.yandex-team.ru/cloud/mdb/mdb-internal-api/internal/models/clusters"
)

var (
	mapDayStringToISOWeekday = map[string]int{
		"MON": 1,
		"TUE": 2,
		"WED": 3,
		"THU": 4,
		"FRI": 5,
		"SAT": 6,
		"SUN": 7,
	}
)

func GetNewMaintenanceTime(maintenanceInfo clusters.MaintenanceInfo, rescheduleType clusters.RescheduleType, delayedUntil optional.Time) (time.Time, error) {
	var err error
	if !maintenanceInfo.Operation.Valid() {
		return time.Time{}, semerr.InvalidInput("There is no maintenance operation at this time")
	}
	newMaintenanceTime := maintenanceInfo.Operation.DelayedUntil
	switch rescheduleType {
	case clusters.RescheduleTypeUnspecified:
		return time.Time{}, semerr.InvalidInput("Reschedule type need to be specified")
	case clusters.RescheduleTypeImmediate:
		newMaintenanceTime = time.Now()
	case clusters.RescheduleTypeSpecificTime:
		newMaintenanceTime, err = delayedUntil.Get()
		if err != nil {
			return time.Time{}, semerr.InvalidInput("Delayed until time need to be specified with specific time rescheduling type")
		}
	case clusters.RescheduleTypeNextAvailableWindow:
		if maintenanceInfo.Window.Anytime() {
			return time.Time{}, semerr.InvalidInput("The window need to be specified with next available window rescheduling type")
		}
		newMaintenanceTime = *CalculateNearestMaintenanceWindow(
			maintenanceInfo.Operation.DelayedUntil,
			maintenanceInfo.Window,
		)
	}
	if newMaintenanceTime == maintenanceInfo.Operation.DelayedUntil {
		return time.Time{}, semerr.InvalidInput("No changes detected")
	}
	if newMaintenanceTime.After(maintenanceInfo.Operation.LatestMaintenanceTime) &&
		rescheduleType != clusters.RescheduleTypeImmediate {
		return time.Time{}, semerr.InvalidInput(fmt.Sprintf("Maintenance operation time cannot exceed %s", maintenanceInfo.Operation.LatestMaintenanceTime.Format(time.RFC3339)))
	}

	return newMaintenanceTime, nil
}

func CalculateNearestMaintenanceWindow(operationTime time.Time, maintenanceWindow clusters.MaintenanceWindow) *time.Time {
	if maintenanceWindow.Anytime() {
		return nil
	}

	operationWeekday := mapDayStringToISOWeekday[strings.ToUpper(operationTime.Weekday().String()[:3])]
	windowWeekday := mapDayStringToISOWeekday[maintenanceWindow.Day]

	res := operationTime
	if operationWeekday == windowWeekday {
		if operationTime.Hour() >= maintenanceWindow.Hour-1 {
			res = res.AddDate(0, 0, 7)
		}
	} else if operationWeekday < windowWeekday {
		res = res.AddDate(0, 0, windowWeekday-operationWeekday)
	} else {
		res = res.AddDate(0, 0, 7+windowWeekday-operationWeekday)
	}

	res = time.Date(res.Year(), res.Month(), res.Day(), maintenanceWindow.Hour-1, res.Minute(), res.Second(), res.Nanosecond(), res.Location())

	return &res
}

func CalculateLatestMaintenanceTime(createdAt time.Time) time.Time {
	return createdAt.AddDate(0, 0, 7*3) // Add 3 weeks
}
